File: rust/src/problems/p53.rs
pub struct Solution;

impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut dp = nums[0];
        let mut result = dp;
        for &n in &nums[1..] {
            dp = n.max(n + dp);
            result = result.max(dp);
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn case1() {
        let nums = vec![-2, 1, -3, 4, -1, 2, 1, -5, 4];
        assert_eq!(6, Solution::max_sub_array(nums));
    }

    #[test]
    fn case2() {
        let nums = vec![-100000];
        assert_eq!(-100000, Solution::max_sub_array(nums));
    }
}

File: cpp/p53.cpp
#include <cassert>
#include <climits>
#include <vector>
using namespace std;

class Solution {
public:
    int maxSubArray(vector<int> &nums) {
        int f = nums[0], ans = nums[0];
        for (int i = 1; i < nums.size(); i++) {
            f = max(nums[i], f + nums[i]);
            ans = max(ans, f);
        }
        return ans;
    }
};

int main() {
    vector<int> nums = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    int ans = Solution().maxSubArray(nums);
    assert(6 == ans);
    return 0;
}